recipe-client-addon/lib/ShieldPreferences.jsm

```diff
@@ -0,0 +1,119 @@
+      Services.obs.addObserver(this, "advanced-pane-loaded");
+      CleanupManager.addCleanupHandler(() => {
+        Services.obs.removeObserver(this, "advanced-pane-loaded");
+      });
+    }
+  },
+
+  observe(subject, topic, data) {
+    switch (topic) {
+      // Add the opt-out-study checkbox to the Privacy preferences when it is shown.
+      case "advanced-pane-loaded":
+        if (!Services.prefs.getBoolPref("browser.preferences.useOldOrganization", false)) {
+          this.injectOptOutStudyCheckbox(subject.document);
+        }
+        break;
+      // If the FHR pref changes, set the opt-out-study pref to the value it is changing to.
+      case NS_PREFBRANCH_PREFCHANGE_TOPIC_ID:
+        if (data === FHR_UPLOAD_ENABLED_PREF) {
+          const fhrUploadEnabled = Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF);
+          Services.prefs.setBoolPref(OPT_OUT_STUDIES_ENABLED_PREF, fhrUploadEnabled);
+        }
+        break;
+    }
+  },
+
+  /**
+   * Injects the opt-out-study preference checkbox into about:preferences and
+   * handles events coming from the UI for it.
+   */
+  injectOptOutStudyCheckbox(doc) {
+    const container = doc.createElementNS(XUL_NS, "vbox");
+    container.classList.add("indent");
+
+    const hContainer = doc.createElementNS(XUL_NS, "hbox");
+    hContainer.setAttribute("align", "center");
+    container.appendChild(hContainer);
+
+    const checkbox = doc.createElementNS(XUL_NS, "checkbox");
+    checkbox.setAttribute("id", "optOutStudiesEnabled");
+    checkbox.setAttribute("label", "Allow Firefox to install and run studies");
+    checkbox.setAttribute("preference", OPT_OUT_STUDIES_ENABLED_PREF);
+    checkbox.setAttribute("disabled", !Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF));
+    hContainer.appendChild(checkbox);
+
+    const viewStudies = doc.createElementNS(XUL_NS, "label");
+    viewStudies.setAttribute("id", "viewShieldStudies");
+    viewStudies.setAttribute("href", "about:studies");
+    viewStudies.setAttribute("useoriginprincipal", true);
+    viewStudies.textContent = "View Firefox Studies";
+    viewStudies.classList.add("learnMore", "text-link");
+    hContainer.appendChild(viewStudies);
+
+    // <prefrence> elements for prefs that we need to monitor while the page is open.
+    const optOutPref = doc.createElementNS(XUL_NS, "preference");
+    optOutPref.setAttribute("id", OPT_OUT_STUDIES_ENABLED_PREF);
+    optOutPref.setAttribute("name", OPT_OUT_STUDIES_ENABLED_PREF);
+    optOutPref.setAttribute("type", "bool");
+
+    // Weirdly, FHR doesn't have a <preference> element on the page, so we create it.
+    const fhrPref = doc.createElementNS(XUL_NS, "preference");
+    fhrPref.setAttribute("id", FHR_UPLOAD_ENABLED_PREF);
+    fhrPref.setAttribute("name", FHR_UPLOAD_ENABLED_PREF);
+    fhrPref.setAttribute("type", "bool");
+    fhrPref.addEventListener("change", () => {
+      checkbox.disabled = !Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF);
```

Comment:
Oh, OK, I hadn't looked at more context, sorry. So, yeah, this is about lifetimes of things. You're right that the JSM should stay loaded pretty much ~forever (though it would get unloaded if the add-on is disabled). The worry I had originally was that the JSM would keep references to the document, thus meaning the tab's document (and with that likely the entire parent tree ie browser window) would live as long as the jsm would live, because the jsm had permanent-ish references to the document (or ones that were cyclical via event handlers). That's now no longer the case, so I think this is fine and shouldn't leak. The document has references to the JSM, but that's fine because it shouldn't prevent the document from being GC'd, because although the document can reach the JSM, the JSM can't reach the document.